c-dependencies/js-compute-runtime/js-compute-builtins.cpp

```diff
@@ -1653,7 +1654,26 @@ bool body_reader_then_handler(JSContext *cx, HandleObject body_owner, HandleValu
     if (Response::is_instance(body_owner)) {
       FetchEvent::set_state(FetchEvent::instance(), FetchEvent::State::responseDone);
     }
-    return HANDLE_RESULT(cx, xqd_body_close(body_handle));
+
+    if (!HANDLE_RESULT(cx, xqd_body_close(body_handle))) {
+      return false;
+    }
+
+    // If the owner is a Request, it's now safe to push it into the pending requests queue as its
+    // body is closed.
+    RootedValue ownerVal{cx};
+    if (!JS_GetProperty(cx, closure, "owner", &ownerVal)) {
+      return false;
+    }
+
+    RootedObject owner{cx, &ownerVal.toObject()};
+    if (Request::is_instance(owner)) {
+      if (!pending_requests->append(owner)) {
+        return false;
+      }
+    }
```

Comment:
This is part 2 from the PR description. Note that since `body_reader_then_handler` is used for both request and response body processing, we only enqueue when the owner of the stream is a `Request`.

Reply:
Is `owner` different than `body_owner`? I'm not following the types real well here but I would have guessed that if `RequestOrResponse::body_handle` is valid to call on `body_owner`, and so is `Response::is_instance`, then surely `Request::is_instance` is too?

If so then I'd expect to see something more like this here:

```c++
    if (Response::is_instance(body_owner)) {
      FetchEvent::set_state(FetchEvent::instance(), FetchEvent::State::responseDone);
    } else {
      if (!pending_requests->append(body_owner)) {
        return false;
      }
    }
```

(Optionally, with an `assert(Request::is_instance(body_owner))` in the `else` branch.)

And in that case, does that mean you can undo the introduction of the `closure` object and just go back to passing the `reader` around?

Reply:
You're totally right, `body_owner` will be baked in as the receiver for the handler, and will be either the `Response` or `Request` objects. I'll revert the introduction of the `closure`, thanks for catching this!